lib/web/device_trust_test.go

```diff
@@ -36,85 +36,105 @@ import (
 func TestHandler_DeviceWebConfirm(t *testing.T) {
 	t.Parallel()
 
-	fakeDevices := &fakeDevicesClient{}
-	wPack := newWebPack(
-		t,
-		1, /* numProxies */
-		withDevicesClientOverride(fakeDevices),
-	)
-
-	proxy := wPack.proxies[0]
-	aPack := proxy.authPack(t, "llama", nil /* roles */)
-	webClient := aPack.clt
-
-	ctx := context.Background()
-
-	t.Run("ok", func(t *testing.T) {
-		query := make(url.Values)
-		query.Set("id", "my-token-id")
-		query.Set("token", "my-token-token")
-
-		// Detect client redirects.
-		var redirected bool
-		httpClient := webClient.HTTPClient()
-		httpClient.CheckRedirect = func(req *http.Request, via []*http.Request) error {
-			// Ignore any subsequent redirects that may happen.
-			if redirected {
-				return nil
+	tests := []struct {
+		name               string
+		redirectURI        string
+		expectedRedirectTo string
+	}{
+		{
+			name:               "no redirect_uri",
+			redirectURI:        "",
+			expectedRedirectTo: "/web",
+		},
+		{
+			name:               "with redirect_uri",
+			redirectURI:        "https://example.com/web/custom/path",
+			expectedRedirectTo: "/web/custom/path",
+		},
+		{
+			name:               "with invalid redirect_uri",
+			redirectURI:        "://invalid",
+			expectedRedirectTo: "/web",
+		},
+		{
+			name:               "with external redirect_uri",
+			redirectURI:        "https://example.com/path",
+			expectedRedirectTo: "/web/path",
+		},
+		{
+			name:               "with empty path redirect_uri",
+			redirectURI:        "https://example.com",
```

Comment:
What happens when the initial URL is `https://myapp.example.com` ?

Reply:
When a user requests an app url and they aren't authenticated, they get redirected with a redirect uri of `https://example.com/web/launch/myapp.example.com` so this is caught the same as the other URLs. but I will add a test for that anyways. Thank you!